src/backend/tty.rs

```diff
@@ -1282,6 +1288,54 @@ impl Tty {
             return;
         };
 
+        if let Some(vblank_throttle_timer_token) = output_state.vblank_throttle_timer_token.take() {
+            niri.event_loop.remove(vblank_throttle_timer_token);
+        }
+
+        let vblank_remaining_time =
+            output_state
+                .last_drm_vblank_timestamp
+                .and_then(|last_drm_vblank_timestamp| {
+                    let refresh_interval = output_state.frame_clock.refresh_interval()?;
+                    let vblank_diff = time.saturating_sub(last_drm_vblank_timestamp);
+                    Some(refresh_interval.saturating_sub(vblank_diff))
+                });
+
+        if let Some(vblank_remaining_time) = vblank_remaining_time.filter(|vblank_remaining_time| {
+            output_state
+                .frame_clock
+                .refresh_interval()
+                .map(|refresh_interval| *vblank_remaining_time > refresh_interval / 2)
+                .unwrap_or(false)
+        }) {
+            static WARN_ONCE: Once = Once::new();
+            WARN_ONCE.call_once(|| warn!("output running faster as expected, throttling vblanks"));
+
+            let throttled_time = if presentation_time.is_zero() {
+                Duration::ZERO
+            } else {
+                presentation_time.saturating_add(vblank_remaining_time)
+            };
+            let throttled_metadata = DrmEventMetadata {
+                sequence: meta.sequence,
+                time: DrmEventTime::Monotonic(throttled_time),
```

Comment:
Wouldn't it be better to set `DrmEventTime::Monotonic(Duration::ZERO)` here? So that it doesn't get set as as HwClock presentation.